Async let second review (#1409)
apple/swift-evolution: proposals/0317-async-let.md

### Requiring an `await`on any execution path that waits for an `async let`

In initial versions of this proposal, we considered a rule to force an `async let` declaration to be awaited on each control-flow path that the execution of a function might take. This rule turned out to be too simplistic, because it doesn't account for the places in which implicit suspension points actually occur. For example, consider a function with an `async let` in a loop:

```swift
func runLoop() async {
  for e in list {
    async let a = f(e)
    guard <condition> else {

>       break // cancels and implicitly awaits the task that produces a

Reviewer: ```suggestion
      break // cancels and implicitly awaits the task that produces "a"
```

    }
    ... await a ...
  }
  foo()
}
```

or a function where the implicit await occurs due to an error being both thrown and caught:

```swift
func runException() async {
  do {
    async let a = f()

>     try mayFail() // cancels and implicitly awaits the task that produces a

Reviewer: ```suggestion
    try mayFail() // cancels and implicitly awaits the task that produces "a"
```

    ... await a ...
  } catch {
    ...
  }
}
```

> In the second example, the `do` block will exit when `mayFail()` throws an error, at which point the child task that computes `a` will need to be cancelled and awaited. However, there is no place to write `await a` to make that suspension point implicit. Therefore, requiring every potential suspension point due to an `async let` to be explicitly marked will require an extension to the Swift grammar.

Reviewer: ```suggestion
In the second example, the `do` block will exit when `mayFail()` throws an error, at which point the child task that computes `a` will need to be cancelled and awaited. However, there is no place to write `await a` to make that suspension point explicit. Therefore, requiring every potential suspension point due to an `async let` to be explicitly marked will require an extension to the Swift grammar.
```

The most promising approach to marking all `async let` suspension points explicitly involves marking the control-flow edges that can result in a potential suspension point with `await`. For the first example, this means using `await break`:

```swift
func runLoop() async {
  for e in list {
    async let a = f(e)
    guard <condition> else {

>       await break   // awaits the child task that produces the value a

Reviewer: ```suggestion
      await break   // awaits the child task that produces the value "a"
```

```swift
func runThrow() async {
  do {
    async let a = f()
    if <condition> {
      await throw SomeError() // awaits the child task that produces a
    }
    ... await a ...
  } catch {
    ...
  }
}
```

However, not all control-flow edges involving implicit `async let` suspension points have a specific keyword to which we can attach `await`, because some come from fall-through to subsequent code. For such cases, one could have a standalone `await` statement marking that fall through:

```swift
func runIfFallthrough() async {
  if <condition> {
    async let a = f()
    ... code ...
    // falling out of this block must await the child task that produces a, so require a freestanding "await"
    await
  }
  ... more code ...
}
```

The same would be required in, e.g., the cases of a `switch` statement that introduce an `async let`:

```swift
func runSwitchCase() async {
  switch <expression> {
  case .a:
    async let a = f()
    // falling out of this block must await the child task that produces a, so require a freestanding "await"
    await

  default:
    ... code ...
  }
  ... code ...
}
```

> The above is a significant expansion of the grammar: introducing the `await` keyword in front of `break`, `continue`, `throw`, and `fallthrough`; requiring `await` on certain throwing expressions; and adding the freestanding `await`statement. It would also need to be coupled with rules that only require the new `await` when it is semantically meaningful. For example, the additional `await` shouldn't be required if all of the `async let` child tasks have already been explicitly awaited in some other manner, e.g.,

Reviewer: ```suggestion
The above is a significant expansion of the grammar: introducing the `await` keyword in front of `break`, `continue`, `throw`, and `fallthrough`; requiring `await` on certain throwing expressions; and adding the freestanding `await` statement. It would also need to be coupled with rules that only require the new `await` when it is semantically meaningful. For example, the additional `await` shouldn't be required if all of the `async let` child tasks have already been explicitly awaited in some other manner, e.g.,
```

```swift
async let veggies = chopVegetables()
...
await veggies
```

### Braces around the `async let` initializer

The expression on the right-hand side of an `async let` declaration is executed in a separate, child task that is running concurrently with the function that initiates the `async let`. It has been suggested that the task should be called out more explicitly by adding a separate set of braces around the expression, e.g.,

```swift
async let veggies = { try await chopVegetables() }
```

The problem with requiring braces is that it breaks the equivalence between the type of the entity being declared (`veggies` is of type `[Vegetable]`) and the value it is initialized with (which now appears to be `@Sendable () async throws -> [Vegetable]`). This equivalence holds throughout nearly all of the language; the only real exception is the `if let` syntax, which which strips a level of optionality and is often considered a design mistake in Swift. For `async let`, requiring the braces would become particularly awkward if one were defining a value of closure type:

```swift
async let closure = { { try await getClosure() } }
```

> Requiring braces on the right-hand side of `async let` would be a departure from Swift's existing precedent with `let` declarations. In the cases where one is defining a longer child task, it is reasonable to create and immediately call a closure, which is common practice with `lazy` variables:

Reviewer: ```suggestion
Requiring braces on the right-hand side of `async let` would be a departure from Swift's existing precedent with `let` declarations. In the cases where one is defining a syntactically larger child task, it is reasonable to create and immediately call a closure, which is common practice with `lazy` variables:
```
